File: benchmarks/lua/compute/matrix.lua
-- Matrix multiplication benchmark
-- Multiply two NxN matrices using table-of-tables representation
-- Tests nested table access + numerical computation

local N = tonumber(arg and arg[1]) or 200
collectgarbage("collect")

local function make_matrix(rows, cols, init_fn)
    local m = {}
    for i = 1, rows do
        m[i] = {}
        for j = 1, cols do
            m[i][j] = init_fn(i, j)
        end
    end
    return m
end

local function multiply(a, b, n)
    local c = {}
    for i = 1, n do
        c[i] = {}
        for j = 1, n do
            local sum = 0.0
            for k = 1, n do
                sum = sum + a[i][k] * b[k][j]
            end
            c[i][j] = sum
        end
    end
    return c
end

-- Initialize matrices
local A = make_matrix(N, N, function(i, j) return (i + j - 1) % 7 + 1.0 end)
local B = make_matrix(N, N, function(i, j) return (i * j) % 11 + 1.0 end)

-- Multiply 3 times: C = A*B, D = C*B, E = D*B
local C = multiply(A, B, N)
local D = multiply(C, B, N)
local E = multiply(D, B, N)

-- Checksum: sum of corner elements
local checksum = E[1][1] + E[1][N] + E[N][1] + E[N][N]
print(string.format("%.6f", checksum))

File: benchmarks/js/compute/matrix.js
// Matrix multiplication benchmark
// Multiply two NxN matrices using array-of-arrays representation
// Tests nested array access + numerical computation

var print = typeof console !== 'undefined' ? console.log.bind(console) : globalThis.print;
var DEFAULT = 200;
var N = parseInt(typeof scriptArgs !== 'undefined' ? scriptArgs[1] : (typeof process !== 'undefined' ? process.argv[2] : '0')) || DEFAULT;

function make_matrix(rows, cols, init_fn) {
    var m = new Array(rows);
    for (var i = 0; i < rows; i++) {
        m[i] = new Array(cols);
        for (var j = 0; j < cols; j++) {
            m[i][j] = init_fn(i + 1, j + 1);
        }
    }
    return m;
}

function multiply(a, b, n) {
    var c = new Array(n);
    for (var i = 0; i < n; i++) {
        c[i] = new Array(n);
        for (var j = 0; j < n; j++) {
            var sum = 0.0;
            for (var k = 0; k < n; k++) {
                sum += a[i][k] * b[k][j];
            }
            c[i][j] = sum;
        }
    }
    return c;
}

// Initialize matrices (using 1-based logic for same values as Lua)
var A = make_matrix(N, N, function(i, j) { return (i + j - 1) % 7 + 1.0; });
var B = make_matrix(N, N, function(i, j) { return (i * j) % 11 + 1.0; });

// Multiply 3 times: C = A*B, D = C*B, E = D*B
var C = multiply(A, B, N);
var D = multiply(C, B, N);
var E = multiply(D, B, N);

// Checksum: sum of corner elements (0-indexed)
var checksum = E[0][0] + E[0][N-1] + E[N-1][0] + E[N-1][N-1];
print(checksum.toFixed(6));
